src/query_graph/mod.rs

```diff
@@ -504,4 +567,38 @@ impl QueryGraph {
             }
         };
     }
+
+    pub(crate) fn get_locally_satisfiable_key(
+        &self,
+        _node: NodeIndex,
+    ) -> Result<Option<NormalizedSelectionSet>, FederationError> {
+        todo!()
+    }
+
+    pub(crate) fn is_cross_subgraph_edge(&self, edge: EdgeIndex) -> Result<bool, FederationError> {
+        let (head, tail) = self.edge_endpoints(edge)?;
+        let head_weight = self.node_weight(head)?;
+        let tail_weight = self.node_weight(tail)?;
+        Ok(head_weight.source != tail_weight.source)
+    }
+
+    pub(crate) fn is_provides_edge(&self, edge: EdgeIndex) -> Result<bool, FederationError> {
+        let edge_weight = self.edge_weight(edge)?;
+        let QueryGraphEdgeTransition::FieldCollection {
+            is_part_of_provides,
+            ..
+        } = &edge_weight.transition
+        else {
+            return Ok(false);
+        };
+        Ok(*is_part_of_provides)
+    }
+
+    pub(crate) fn has_an_implementation_with_provides(
+        &self,
+        _source: &NodeStr,
+        _interface_field_definition_position: InterfaceFieldDefinitionPosition,
+    ) -> Result<bool, FederationError> {
+        todo!()
```

Comment:
we’ll definitely need to grep the entire code base for `FIXME`/`TODO`/`todo!` at some point